### PR TITLE
[MoE] Moe apis

### PR DESCRIPTION
**Pull request by sljlp**
### PR types

New features
### PR changes

APIs
### Describe

add apis about MoE

### CODE REVIEW

**Review by ForFishes — APPROVED** [review]
LGTM

**Review by wanghuancoder — APPROVED** [review]

File: python/paddle/incubate/distributed/models/moe/moe_layer.py [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+        fwd_batch_size = int(fwd_expert_count.sum().item())
+    return (
+        pos,
+        local_expert_count,
+        global_expert_count,
+        fwd_expert_count,
+        fwd_batch_size, )
+
+
+class MoeLayer(nn.Layer):
```
Comment (XiaoguangHu01) [inline python/paddle/incubate/distributed/models/moe/moe_layer.py:228]: MoeLayer和MOEGather缩写方式不一致，建议统一采用业界通用的写法。

**Review by XiaoguangHu01 — APPROVED** [review]
LG API

File: python/paddle/incubate/distributed/models/moe/moe_layer.py (hunk 1 above)
Comment (sljlp) [reply python/paddle/incubate/distributed/models/moe/moe_layer.py:228]: OK

File: python/paddle/incubate/distributed/models/moe/gate/__init__.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Ligoml) [inline python/paddle/incubate/distributed/models/moe/gate/__init__.py:1]: Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.

File: python/paddle/incubate/distributed/models/moe/gate/base_gate.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Ligoml) [inline python/paddle/incubate/distributed/models/moe/gate/base_gate.py:1]: Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.

File: python/paddle/incubate/distributed/models/moe/gate/gshard_gate.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Ligoml) [inline python/paddle/incubate/distributed/models/moe/gate/gshard_gate.py:1]: Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.

File: python/paddle/incubate/distributed/models/moe/gate/naive_gate.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Ligoml) [inline python/paddle/incubate/distributed/models/moe/gate/naive_gate.py:1]: Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.

File: python/paddle/incubate/distributed/models/moe/gate/switch_gate.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Ligoml) [inline python/paddle/incubate/distributed/models/moe/gate/switch_gate.py:1]: Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.

File: python/paddle/incubate/distributed/models/moe/grad_clip.py [hunk 7]
```diff
@@ -0,0 +1,217 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Ligoml) [inline python/paddle/incubate/distributed/models/moe/grad_clip.py:1]: Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.

File: python/paddle/incubate/distributed/models/moe/moe_layer.py [hunk 8]
```diff
@@ -0,0 +1,431 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Ligoml) [inline python/paddle/incubate/distributed/models/moe/moe_layer.py:1]: Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.

File: python/paddle/incubate/distributed/models/moe/moe_layer.py [hunk 9]
```diff
@@ -0,0 +1,431 @@
...
+    return (
+        pos,
+        local_expert_count,
+        global_expert_count,
+        fwd_expert_count,
+        fwd_batch_size, )
+
+
+class MoeLayer(nn.Layer):
+    """Moe Layer
```
Comment (Ligoml) [inline python/paddle/incubate/distributed/models/moe/moe_layer.py:229]: 缺失对api的定义与功能描述，Moe Layer是什么，什么场景下做什么用的？

File: python/paddle/incubate/distributed/models/moe/moe_layer.py [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+    Args:
+        d_model: (int) model dimention
+        experts: (nn.LayerList) expert networks list
+        gate: (dict|NaiveGate|SwitchGate|NaiveGate):
```
Comment (Ligoml) [inline python/paddle/incubate/distributed/models/moe/moe_layer.py:233]: 参考格式：

d_model (int, optional): model dimention.
experts (nn.LayerList, optional): expert networks list.
gate (dict|NaiveGate|SwitchGate|NaiveGate): 

另外对参数的介绍可以再清晰一些

File: python/paddle/incubate/distributed/models/moe/moe_layer.py [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+        experts: (nn.LayerList) expert networks list
+        gate: (dict|NaiveGate|SwitchGate|NaiveGate): 
+                if gate is a dict:
+                    gate is a gate network config, containing 2 keys: 
+                    `type`(str) value can be: "naive", "gshard", "switch" or None, default is "gshard"
+                    `top_k`(int) default value is 2
+                else gate is an instance of NaiveGate|SwitchGate|NaiveGate:
+
+        moe_group: moe group for experts communication
+        mp_group: mp group for mp commutication
```
Comment (Ligoml) [inline python/paddle/incubate/distributed/models/moe/moe_layer.py:241]: 这几个有默认值的参数需要增加default，参考格式：

moe_group : moe group for experts communication. Default: None.
mp_group : mp group for mp commuticatio. Default: None.

File: python/paddle/incubate/distributed/models/moe/moe_layer.py [hunk 12]
```diff
@@ -0,0 +1,431 @@
...
+        kwargs: other parameters
+    Examples:
+        .. code-block:: python
```
Comment (Ligoml) [inline python/paddle/incubate/distributed/models/moe/moe_layer.py:244]: 为了保证正常的样式显示，需要在examples和code之前分别空一行，参考格式：

        kwargs: other parameters

    Examples:
        .. code-block:: python

        from paddle.nn import layer, LayerList
        from paddle.distributed.moe import Moelayer
        from paddle.distributed.collective import Group
        from paddle.distributed import fleet

File: python/paddle/incubate/distributed/models/moe/moe_layer.py [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+
+from paddle.distributed.fleet.meta_parallel import get_rng_state_tracker
+from paddle.distributed import fleet
+from paddle.autograd import PyLayer
+from .gate import NaiveGate, GShardGate, SwitchGate, BaseGate
+from .utils import count_by_gate
+from paddle.distributed.fleet.meta_parallel.pp_utils.utils import _hp_recompute
+from paddle import fluid
+
+__all__ = ["MoeLayer"]
```
Comment (Ligoml) [inline python/paddle/incubate/distributed/models/moe/moe_layer.py:33]: 对外暴露的API需要构建官网文档预览，保证文档正确显示，需要准备对应的中文文档

File: python/paddle/incubate/distributed/models/moe/moe_layer.py [hunk 14]
```diff
@@ -0,0 +1,431 @@
...
+                if gate is a dict:
+                    gate is a gate network config, containing 2 keys: 
+                    `type`(str) value can be: "naive", "gshard", "switch" or None, default is "gshard"
+                    `top_k`(int) default value is 2
+                else gate is an instance of NaiveGate|SwitchGate|NaiveGate:
+
+        moe_group: moe group for experts communication
+        mp_group: mp group for mp commutication
+        kwargs: other parameters
+    Examples:
```
Comment (Ligoml) [inline python/paddle/incubate/distributed/models/moe/moe_layer.py:243]: 在Args和Examples之间缺乏Returns

File: python/paddle/fluid/tests/unittests/test_random_routing_op.py [hunk 15]
```diff
@@ -0,0 +1,77 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Ligoml) [inline python/paddle/fluid/tests/unittests/test_random_routing_op.py:1]: Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.

File: paddle/fluid/operators/prune_gate_by_capacity_op.h [hunk 16]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Ligoml) [inline paddle/fluid/operators/prune_gate_by_capacity_op.h:1]: Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.

File: paddle/fluid/operators/prune_gate_by_capacity_op.cu [hunk 17]
```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Ligoml) [inline paddle/fluid/operators/prune_gate_by_capacity_op.cu:1]: Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.

File: paddle/fluid/operators/prune_gate_by_capacity_op.cc [hunk 18]
```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Ligoml) [inline paddle/fluid/operators/prune_gate_by_capacity_op.cc:1]: Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.

Comment (sljlp) [reply paddle/fluid/operators/prune_gate_by_capacity_op.cc:1]: 好的，下个PR会把问题全部修改掉

**Review by Ligoml — APPROVED** [review]

**Pull request merged**